### PR TITLE
chore: move `delayed_type_checks` field from `NodeInterner` to `TypeChecker`

### PR DESCRIPTION
**Pull request by f01dab1e**
## Problem

The current implementation of the node interner in the codebase appears to be handling multiple responsibilities and exhibiting signs of complexity. 

# PR Checklist

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Review by jfecher — APPROVED** [review]
Looks good! I'm not sure why we didn't do it this way to begin with.

**Comment by f01dab1e** [comment]
Hey, @jfecher! How do you think there is a sense to move fields into separate structures: 
- `next_type_variable_id`
- `globals`
- `primitive_methods`, `struct_methods`

**Comment by jfecher** [comment]
In general anything that is used across both name resolution and type inference (+ potentially monomorphization) belongs in the node interner. Since the interner is the general context object that accompanies the HIR.
- `next_type_variable_id` - also referenced by name resolution when translating types, so it is fine in the node interner.
- `globals` - I believe these are referenced in name resolution until they're later inlined during monomorphization. So they should be fine as well.
- `primitive_methods`/`struct_methods` - they're populated during name resolution and resolved during type checking so the node interner is fine for these as well.

So in summary, I think each of these are fine where they are.
